[llvm][ARM] Emit MVE .arch_extension after .fpu directive if it does not include MVE features

simpal01 opened the pull request:
The floating-point and MVE features together specify the MVE functionality that is supported on the Cortex-M85 processor. But the FPU extension for the underlying architecture(armv8.1-m.main) is FPV5 which does not include MVE-F. So Compiler's -S output and `-save-temps=obj` loses MVE feature which leads to assembler error. What happening here is .fpu directive overrides any previously set features by .cpu directive. Since the the corresponding .fpu generated (.fpu fpv5-d16) does not include MVE-F, it overrides those features even though it is supported and set by the .cpu directive. Looks like .fpu is supposed to do this.

In this case, there should be an .arch_extension directive re-enabling the relevant extensions after .fpu  if the goal is to keep these extensions enabled. GCC also does the same.

So this patch enables the MVE features by emitting the below arch extension:
  .fpu fpv5-d16
  .arch_extension mve.fp

Commit:
[ARM] .fpu equals fpv5-d16 disables floating point MVE which leads to unsupported MVE instructions for cortex M85/M55.

The floating-point and MVE features together specify the MVE functionality that is supported on the Cortex-M85 processor. But the FPU extension for the underlying architecture(armv8.1-m.main) is FPV5 which does not include MVE-F.

So either when we explictly specify -mfpu=fpv5-d16 or Compiler's -S output and `-save-temps=obj` loses MVE feature which leads to assembler error. What happening here is .fpu directive overrides any previously set features by .cpu directive. Since the the corresponding .fpu generated (.fpu fpv5-d16) does not include MVE-F, it overrides those features even though it is supported and set by the .cpu directive. Looks like .fpu is supposed to do this.

In this case, there should be an .arch_extension directive re-enabling the relevant extensions after .fpu  if the goal is to keep these extensions enabled. GCC also does the same.

So this patch enables the MVE features by:
  .fpu fpv5-d16
  .arch_extension mve.fp

DavidSpickett commented:
I think the commit title would make more sense at a glance if it was saying what is being done instead of what's being fixed. How about:
```
[llvm][ARM] Emit MVE .arch_extension after .fpu directive if it does not include MVE features
```
Then the commit message is fine as is, and tells the reader why this was done and what it fixes.

DavidSpickett commented on clang/test/CodeGen/arm-v8.1m-check-mve.ll:
```diff
@@ -0,0 +1,56 @@
+  %18 = load float, ptr %d7, align 4
+  %19 = call float @llvm.fmuladd.f32(float 0x3FA47AE140000000, float %sub6, float %18)
+  store float %19, ptr %d7, align 4
+  ret i8 0
```
Do you need such a large example just to get the directives emitted?

DavidSpickett replied:
Ok I see, you need to setup the call to this fmuladd, that you know will emit MVE instructions.

In that case, could you add a check line for one of the MVE instructions as well? That would make the intent clearer, and warn us in the (unlikely) case that this code stopped emitting MVE instructions.

DavidSpickett commented on clang/test/CodeGen/arm-v8.1m-check-mve.ll:
```diff
@@ -0,0 +1,56 @@
+; RUN: %clang --target=arm-none-eabi -mcpu=cortex-m85 -mfloat-abi=hard -O2 -c -mthumb -save-temps=obj %s
+; RUN: %clang --target=arm-none-eabi -mcpu=cortex-m55 -mfloat-abi=hard -O2 -c -mthumb -save-temps=obj %s
+; CHECK: .fpu   fpv5-d16
+; CHECK  .arch_extension mve.fp
```
Is arch_extension expected to always be on the very next line? If so, CHECK-NEXT would be bit more robust.

simpal01 replied:
Done

DavidSpickett commented on clang/test/CodeGen/arm-v8.1m-check-mve.ll:
```diff
@@ -0,0 +1,56 @@
+; CHECK: .fpu   fpv5-d16
+; CHECK  .arch_extension mve.fp
+target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
+target triple = "thumbv8.1m.main-none-unknown-eabihf"
```
Not 100% sure but I think you can leave these out and they'll be got from the options given to clang regardless (not very important though).

simpal01 replied:
Done

simpal01 commented:
> I think the commit title would make more sense at a glance if it was saying what is being done instead of what's being fixed. How about:
> 
> ```
> [llvm][ARM] Emit MVE .arch_extension after .fpu directive if it does not include MVE features
> ```
> 
> Then the commit message is fine as is, and tells the reader why this was done and what it fixes.

Done

simpal01 commented:
> I think the commit title would make more sense at a glance if it was saying what is being done instead of what's being fixed. How about:
> 
> ```
> [llvm][ARM] Emit MVE .arch_extension after .fpu directive if it does not include MVE features
> ```
> 
> Then the commit message is fine as is, and tells the reader why this was done and what it fixes.

Done

Commit:
fixup! [llvm][ARM] Emit MVE .arch_extension after .fpu directive if it does not include MVE features.

  1. .arch_extension will always be on the very next line.
     CHECK-NEXT would be bit more robust.

Commit:
fixup! [ARM] .fpu equals fpv5-d16 disables floating point MVE which leads to unsupported MVE instructions for cortex M85/M55.

Commit:
fixup! [ARM] .fpu equals fpv5-d16 disables floating point MVE which leads to unsupported MVE instructions for cortex M85/M55.

DavidSpickett commented on clang/test/CodeGen/arm-v8.1m-check-mve.c:
```diff
@@ -0,0 +1,35 @@
+// RUN: %clang --target=arm-none-eabi -mcpu=cortex-m85 -mfloat-abi=hard -O2 -save-temps=obj -S -o - %s | FileCheck %s
```
Should there be an m55 test for this as well, as you've got 55 and 85 for save-temps.

simpal01 replied:
Done

DavidSpickett commented:
Just small things on the test case from me. Nominate @davemgreen to give it a final check and approval, since I have not been around this area in a while.

Commit:
fixup! [ARM] .fpu equals fpv5-d16 disables floating point MVE which leads to unsupported MVE instructions for cortex M85/M55.

simpal01 commented on clang/test/CodeGen/arm-v8.1m-check-mve.c:
```diff
@@ -0,0 +1,35 @@
+#define DUMMY_CONST_2 (0.01F)
+#define DUMMY_CONST_3 (0.02F)
+#define DUMMY_CONST_4 (0.03F)
+#define DUMMY_CONST_5 (0.04F)
```
Done

simpal01 commented on clang/test/CodeGen/arm-v8.1m-check-mve.c:
```diff
@@ -0,0 +1,35 @@
+// RUN: %clang --target=arm-none-eabi -mcpu=cortex-m85 -mfloat-abi=hard -O2 -save-temps=obj -S -o - %s | FileCheck %s
+// RUN: %clang --target=arm-none-eabi -mcpu=cortex-m85 -mfloat-abi=hard -O2 -c -mthumb -save-temps=obj %s
+// RUN: %clang --target=arm-none-eabi -mcpu=cortex-m55 -mfloat-abi=hard -O2 -c -mthumb -save-temps=obj %s
```
Done

DavidSpickett commented:
Tests now LGTM, thanks for the updates.

davemgreen commented on clang/test/CodeGen/arm-v8.1m-check-mve.c:
```diff
@@ -0,0 +1,35 @@
+
+// CHECK-LABEL: foo
+// CHECK: vsub.f32
+// CHECK: vfma.f32
```
This looks like it is trying to SLP vectorize to vector operations? These instructions will match both `vsub.f32 s6, s2, s10` and `vsub.f32 q6, q2, q1` operations though.

davemgreen reviewed (commented):
Hi. From what I can tell the logic looks OK. We add the archextension in places we expect to now?

It is generally considered best practice to not have clang test that run the entire pass pipeline. In this case it looks like it's trying to SLP vectorize the code to make sure MVE operations are produced?  The problem can be that decision like that can change, and it is better if the tests are more narrowly focussed than replying on the whole pipeline. You do loose end-to-end testing but it may be better to have a clang test that checks the IR generated is what is expected, and an assembly test to make sure the `.arch_extension mve.fp` is recognized and turns on the MVE instructions it should do.

Commit:
fixup! [ARM] .fpu equals fpv5-d16 disables floating point MVE which leads to unsupported MVE instructions for cortex M85/M55.

 The tests are now more narrowly focussed than running on the whole pipeline.
 Now added:
   1.  a llc test that checks the assembly generated contains the .arch_extension mve.fp
       if it supports the mve floating point operations.
   2.  an assembly test to make sure the .arch_extension mve.fp is recognized and
       turns on the MVE instructions it should do.

simpal01 commented:
> Hi. From what I can tell the logic looks OK. We add the archextension in places we expect to now?
> 
> It is generally considered best practice to not have clang test that run the entire pass pipeline. In this case it looks like it's trying to SLP vectorize the code to make sure MVE operations are produced? The problem can be that decision like that can change, and it is better if the tests are more narrowly focussed than replying on the whole pipeline. You do loose end-to-end testing but it may be better to have a clang test that checks the IR generated is what is expected, and an assembly test to make sure the `.arch_extension mve.fp` is recognized and turns on the MVE instructions it should do.

DONE

davemgreen reviewed (approved):
Thanks. LGTM
